kvdb/etcd/commit_queue.go

```diff
@@ -36,19 +42,24 @@ func NewCommitQueue(ctx context.Context) *commitQueue {
 		ctx:       ctx,
 		readerMap: make(map[string]int),
 		writerMap: make(map[string]int),
-		queue:     make(chan func(), commitQueueSize),
+		queue:     list.New(),
+		shutdown:  make(chan struct{}),
 	}
+	q.queueCond = sync.NewCond(&q.queueMx)
 
 	// Start the queue consumer loop.
-	q.wg.Add(1)
 	go q.mainLoop()
 
 	return q
 }
 
-// Wait waits for the queue to stop (after the queue context has been canceled).
-func (c *commitQueue) Wait() {
-	c.wg.Wait()
+// Stop signals the queue to stop after the queue context has been canceled and
+// waits until the has stopped.
+func (c *commitQueue) Stop() {
+	// Signal the queue's condition variable to ensure the mainLoop reliably
+	// unblocks to check for the exit condition.
+	c.queueCond.Signal()
```

Comment:
In the past we've found that at times this initial signal gets sort of "swallowed" causing us to add a loop around it like in this areas:
  * https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/mailbox.go#L214
  * https://github.com/lightningnetwork/lnd/blob/master/peer/brontide.go#L1049

Reply:
Don't we also need to cancel the main context, or send on some quit channel here?

Reply:
Ah, I wasn't aware of this. I suggested removing the loop in a previous review cycle :sweat_smile:

Reply:
I tried to look up any references about this, but didn't succeed. Is this a know issue with Go conditional variables? Maybe we did the loops because of other circumstances on our side?

Reply:
>  Is this a know issue with Go conditional variables? Maybe we did the loops because of other circumstances on our side?

Good question...we could just be using the API wrong, and then ended up settling on this workaround as it seemed to make a difference in practice. Otherwise, we would see behavior where things would never shutdown without this type of loop added. I guess I view it as more of a defensive thing, but we should def try to repro it in like golang playground or something.

Reply:
FWIW, doesn't seem to be causing any issues in the `itest` (which are actually _all_ green on this PR!!!), so I guess lets just monitor it and see if it triggers any issue re shutdown.

Reply:
sgtm